### PR TITLE
Add VirtualBox screenshot check to avoid to press enter in terminal

### PR DESCRIPTION
**Pull request by MechantRouquin**
__Your script is awesome !!!__

I've added a feature to __compare screenshot__ from the virtualbox with some images references.
That allow to be more push-button by avoiding to press enter in terminal during script execution.

It may or not interresting you ?

The feature take a screenshot every 30 seconds, convert it in 640x480 resolution and compare it to references images (in images directory). It compare _Language_, _Utilities_ and _Terminal_ windows and validate the step if needed.

Enjoy!

### CODE REVIEW

**Comment by myspaghetti** [comment]
Thanks, great work! I think I'll implement this as an optional feature or perhaps put it in a different branch.

The pull request calls `grep` on line `1508` which isn't included in `coreutils` so it has to be mentioned as a separate dependency or replaced with some built-in function. I'll do a proper review eventually.

File: README.md [hunk 1]
```diff
@@ -61,9 +61,9 @@ The VirtualBox EFI implementation does not properly load the FileVault full disk
 ## Dependencies
 
 All the dependencies should be available through a package manager:  
-`bash` `coreutils` `gzip` `unzip` `wget` `xxd` `dmg2img`  `virtualbox`
+`bash` `coreutils` `gzip` `unzip` `wget` `xxd` `dmg2img`  `virtualbox` `ImageMagick`
```
Comment (myspaghetti) [inline README.md:64]: Should include `grep` as a dependency or remove `grep` from `compare_vb_screen_and_ref()`

Comment (MechantRouquin) [reply README.md:64]: I may use something else compatible with coreutils.

**Comment by MechantRouquin** [comment]
Sorry I'm not familiar with GitHub pull request, do I have to make modification and do a new pull request or I can commit changes ?

**Comment by myspaghetti** [comment]
I am currently doing a full review on the pull request. One issue is the use of different resolutions; this can be resolved by only using the center of the image, for example
```
convert "${vm_name}_screenshot.png" -gravity center -crop 640x480+0+0 +repage -scale '64x64!' "${vm_name}_cropped_scaled.png"
```
I will be submitting a commit today or in a few days which I will merge into master.

**Comment by MechantRouquin** [comment]
I'll be in holidays the next couple of weeks, I let you decide what to do or wait the beginning of september.
Sorry I've _pull request_ a bit late.

**Comment by myspaghetti** [comment]
I am referring to the use of different display aspect ratios and resolutions, which will never match the reference image if they are scaled without being cropped. See for example `1280x800` and `5120x2880`. Apparently even `-gravity center` won't overcome this because the GUI is not centered horizontally and vertically, and the image must be processed further before comparing to any reference image.
1280x800_resized
5120x2880_resized

This can possibly be overcome using `-liquid-rescale`, I will look into it later.

**Comment by myspaghetti** [comment]
Closing current pull request because image recognition fails under non-default resolutions. The ImageMagick branch should:

* have base64-encoded reference images (preferably 64 by 64 pixel optimized PNGs) embedded in the `compare_vm_screen_and_ref` function
* use `-liquid-rescale` or similar ImageMagick option that preserves high-energy elements (like the GUI) and removes low-energy elements (like the background)
* be tested on multiple resolutions and aspect ratios, at least `640x480`, `1280x800`, `5120x2880` but preferably all supported resolutions

**Comment by myspaghetti** [comment]
Even with `-liquid-rescale` the results are not great:
```
./convert.exe \( 1280x800.png -gravity center -crop 640x+0+0 +repage \) -liquid-rescale '640x480!' '12808x800_liquid_rescale.png'
```
1280x800_liquid_rescale

```
./convert.exe \( 5120x2880.png -gravity center -crop 640x+0+0 +repage \) -liquid-rescale '640x480!' '5120_2880_liquid_rescale.png'
```
5120_2880_liquid_rescale

```
compare -metric MAE 1280x800_liquid_rescale.png 5120_2880_liquid_rescale.png null:
6983.61 (0.106563)
```
Which is usable if the threshold is, say, 8000, but I feel like it's still prone to error (haven't tested on all resolutions yet).

I believe `-metric RMSE -subimage-search` will produce better results. I'll keep investigating.

**Comment by MechantRouquin** [comment]
Yes many things to test and improve!  
I'll look at it in september.
Thanks for the reviews.

**Comment by myspaghetti** [comment]
Ended up using Tesseract OCR in [version `0.97.0`](https://github.com/myspaghetti/macos-virtualbox/commit/d6ff74e5744465e43c9ecc36c95d6eeaddb3a6ca). Tesseract is two orders of magnitude faster than `compare` and a lot more reliable.

**Comment by MechantRouquin** [comment]
Nice solution.
I'll check it.

**Pull request closed without merging**